Sort containers in index order

ryanbrainard opened the pull request:
This is an implementation of the idea I submitted in https://github.com/derailed/k9s/issues/2843. While my [follow up questions](https://github.com/derailed/k9s/issues/2843#issuecomment-2310407022) were not answered yet, I took a stab at this with my own preferences how it should work. I'm happy to adjust as needed.

- Went with "Index" for the name this sort order. Used '`IDX` as column header to save horizontal space.
- Showing regular containers as the plain index number. Showing init containers prefixed with a `ⓘ` and then the index number.  I did it this way to keep the visuals simple in the regular case, but mark index when it's in the `InitContainers` array. This also preserves the actual index in the pod spec. Another benefit of using the Unicode characters as prefixes here is that the `ⓘ` sorts first so it visually represents the run order of the init containers that run first.
- Dropped the `INIT` column because it is now redundant with the `ⓘ` marker. This further saves horizontal space.
- Made `IDX` the default sort order for containers. This ensures that the containers are listed in the order defined and run (in the case of init containers), which will improve the default UX.

## Demo

### Pod Spec

Given a pod with these init containers and containers:

```yaml
apiVersion: v1
kind: Pod
metadata:
  name: rainbow
spec:
  initContainers:
    - name: init-container-red
      image: busybox:1.28
      command: ['sh', '-c', "sleep 5"]
    - name: init-container-orange
      image: busybox:1.28
      command: ['sh', '-c', "sleep 5"]
  containers:
    - name: container-yellow
      image: busybox:1.28
      command: ['sh', '-c', "sleep 3600"]
    - name: container-green
      image: busybox:1.28
      command: ['sh', '-c', "sleep 3600"]
    - name: container-blue
      image: busybox:1.28
      command: ['sh', '-c', "sleep 3600"]
```

### k9s Containers View

Here is how it is represented in k9s in the same order it is defined. Note the `IDX` column on the left with the init container with `ⓘ` and their index  and regular containers with just their index.

<img width="2462" alt="2024-09-12_15-43-34" src="https://github.com/user-attachments/assets/253157b3-fd7d-47fa-873a-7a929c39cb30">

ryanbrainard commented:
Just pushed up https://github.com/derailed/k9s/pull/2878/commits/2af813b89f614cd6fd79a6b5582ff4b19be0cbdd and https://github.com/derailed/k9s/pull/2878/commits/0e70b11427e26b05d59f903b56736899c1833e42 to resolve the Codebeat issues where `makeContainerRes{}` and `ContainerRes()` had too many fields and arguments by moving `Container` and `Status` from fields to methods that look up the values with` isInit` and `index` instead of passing them in. This did have a ripple effect everywhere container res was used and makes this PR bigger than I hoped, but does improve the safety a bit and makes it easier to expand to return other things in the future since we have the whole pod available.

derailed commented on internal/dao/container.go:
```diff
@@ -47,11 +46,15 @@ func (c *Container) List(ctx context.Context, _ string) ([]runtime.Object, error
 	res := make([]runtime.Object, 0, len(po.Spec.InitContainers)+len(po.Spec.Containers))
-	for _, co := range po.Spec.InitContainers {
-		res = append(res, makeContainerRes(co, po, cmx[co.Name], true))
+	for i, co := range po.Spec.InitContainers {
```
I think the knowledge of the container type is right here i.e what kind of container are we dealing with? 
aka init, main, ephemeral (let's not forget those!). 
I think it's best to precompute all the necessary bits vs lazy eval when one needs the index or other info out of containerRes.
So I think its best (and less intrusive!) to let containerRes  be dumb but add the differentiation right here at the call site and thus mod the current signature to something like this:
```go
func makeContainerRes(idx string, co v1.Container, po *v1.Pod, cmx *mv1beta1.ContainerMetrics) render.ContainerRes {...}
```

Then at the call site:

```go
res = append(res, makeContainerRes("I"+strconv.Itoa(i+1), co, po, cmx[co.Name]))
```

I also think we should go with non-zero index and perhaps use the scheme I, M, E for init, main and ephemeral.

derailed commented on internal/render/container.go:
```diff
@@ -101,22 +101,22 @@ func (c Container) Render(o interface{}, name string, r *model1.Row) error {
 	}
 
-	cur, res := gatherMetrics(co.Container, co.MX)
+	cur, res := gatherMetrics(co.Container(), co.MX)
```
Making co.Container lazy eval is a killer here as we have to keep recomputing the container we're dealing with which is pretty much static.

derailed commented on internal/render/container.go:
```diff
@@ -101,22 +101,22 @@ func (c Container) Render(o interface{}, name string, r *model1.Row) error {
 	}
 
-	r.ID = co.Container.Name
+	r.ID = co.IndexLabel()
```
Same here we keep recomputing the index.

derailed commented on internal/render/container.go:
```diff
@@ -101,22 +101,22 @@ func (c Container) Render(o interface{}, name string, r *model1.Row) error {
 	r.Fields = model1.Fields{
-		co.Container.Name,
+		co.IndexLabel(),
+		co.Container().Name,
```
We are recomputing the container 5 times vs just calling the member. I don't think we should lazy eval this call.

derailed commented on internal/render/container.go:
```diff
@@ -236,13 +236,23 @@ func probe(p *v1.Probe) string {
+}
+
 // ContainerRes represents a container and its metrics.
 type ContainerRes struct {
```
All that should be needed here is to change IsInit->idx and pass this in from the call site.

derailed commented on internal/render/container.go:
```diff
@@ -254,3 +264,43 @@ func (c ContainerRes) GetObjectKind() schema.ObjectKind {
+	} else {
+		containerType = "\u2800" // blank (sorts last)
+	}
+	return fmt.Sprintf("%s %d", containerType, c.Index)
```
From the user perspective best to use 1 index.

derailed commented on internal/render/container.go:
```diff
@@ -254,3 +264,43 @@ func (c ContainerRes) GetObjectKind() schema.ObjectKind {
+func (c ContainerRes) IndexLabel() string {
+	var containerType string
+	if c.IsInit {
+		containerType = "\u24D8" // encircled i (sorts first)
```
Not keen on using an ascii code to distinguish and sort on init. I think `I` prefix should do.

derailed reviewed (changes requested):
@ryanbrainard Thank you for this PR Ryan!
I think that are a few issues here.  
In summary, I think the actual update is simple if we determine at the call site what kind of container we are dealing with. Altering the model lead you to much more pain than necessary and resorting to lazy eval could account for additional perf implications.

ryanbrainard commented:
@derailed Thanks for the review. I can switch this over to being precomputed. I had something more like that originally and caused Codebeat issues, but I can fully precompute it by removing the isInit arg and likely keep within Codebeat's rules.

ryanbrainard commented:
@derailed I just pushed up the requested changes. Looks like this now:

<img width="2435" alt="2024-09-16_11-44-12" src="https://github.com/user-attachments/assets/98506952-a0aa-4cbd-a0aa-0a1e7c9512b4">

A few things to call out:
 - I added in Ephemeral Containers to the containers rendering as requested, but I did not do this in other places like in the pod rendering where it shows container counts.
 - I also extracted the container statuses because the way it currently works on `master` is looking up by `name`, which could fail if two containers of different types have the same names. The way I changed it should be faster and safer.
 - Codebeat is complaining again. It's complaining about complexity, but IMHO, the way it is with the nested loops is better than duplicating the logic in three separate loops. I'm not going to change things again until I hear back from you if this is he right direction.
 - Note that the order is now `E, I, M`, which is alphabetical, but isn't really in run order since Ephemeral Containers would usually be last. That's probably fine since they are rare and at least Init Containers are before the main Containers, but wanted to at least call it out.

derailed commented on internal/dao/container.go:
```diff
@@ -46,14 +46,47 @@ func (c *Container) List(ctx context.Context, _ string) ([]runtime.Object, error
-	for _, co := range po.Spec.InitContainers {
-		res = append(res, makeContainerRes(co, po, cmx[co.Name], true))
+
+	type containerGroup struct {
```
I don't think we need to introduce these groups. If we modified makeContainerRes signature to take in a kind (I,M,E) and an index should be sufficient. Then getContainerStatus can just look up statuses based on kind and index.
What do you think?

ffais replied:
@derailed It might be useful to add [sidecar containers](https://kubernetes.io/docs/concepts/workloads/pods/sidecar-containers/) (S) to the type list. This would help solve this issue #2912

ryanbrainard replied:
@derailed It could be done that way, but I introduced these groups to avoid having the logic spread across `List()` and `getContainerStatus()` with  `for` loops in each that both need to know about different container types. Since `getContainerStatus()` is only used in this one place, it makes it simpler to have the logic in one place and IMHO is easier to read and extend. I purposely made this new type private to this one method so it's not introducing a new global concept.

@ffais yes, I think that should be do-able. (This is actually a good example of how this is easier to extend by adding a predicate to the group to differentiate between sidecars and other init containers.)

ryanbrainard replied:
@derailed well, it looks like you already pulled in your own way of doing things in https://github.com/derailed/k9s/pull/2955/commits/b70b32c5a80cc96ab01d6f4d84559ec3a9142c36. I'll close this PR and the [issue](https://github.com/derailed/k9s/pull/2878).

derailed reviewed (commented):
@ryanbrainard Thank you for the updates Ryan.
